### ISSUE TITLE
Radio didn't turn off after using the switch

### ISSUE THREAD

**Issue by fraternl** [opentx/opentx#2421]
After flashing the 2015-07-05 firmware and checking some settings I wasn't able to turn off the radio. I waited more than 2 minutes. In fact I'm still waiting whilst writing this message.
It's on "Saving model settings"

I can't say if this is a fluke or if it happens on more occasions.
I never had it before. I do hope it doesn't corrupt my models if I unplug the battery....

I unplugged the battery for 5 seconds, plugged it back in.
I turned the radio on and it immediately went to my model (no splash screen... instant) (which could be very good if you are flying at that moment)
I turned it off using the switch and then on again. it went through the normal startup procedure including splash screen. I turned it off....  This time it turned off normal.

Another on/off procedure went normal.

Maybe a fluke...

Is the instant-on some feature????

**Comment by bsongis** [comment]
Instant-on is a feature, yes! In case your radio reboots when flying, you don't want to be blocked on an EEPROM / THR / SW warning, the SD is not mounted to avoid any problem in this area as well.

**Comment by bsongis** [comment]
Now your problem. In OpenTX 2.1 we use hardware I2C instead of software I2C, it now goes around 5 times faster to read / write the EEPROM. Would you do more tests, you just modify something, shutdown.

**Comment by kilrah** [comment]
I've had mine lock at the same moment once a couple of days ago. Happened only once, but with your report there might be something.

**Comment by bsongis** [comment]
Taranis Plus for both reports?

**Comment by kilrah** [comment]
For me yes

**Comment by RC-SOAR** [comment]
Same issues as the o/p here (need to disconnect battery to shut down), but with an earlier build.

Possibly related to this, I've had a couple of freezes when scrolling up/down the model list. Can't remember the exact details, but I do remember the channels continued to respond correctly.

Tried to reproduce the issue with later builds, but unlike the o/o have not been able to do so. But it's been on my mind.

**Comment by bsongis** [comment]
I succeeded to reproduce it. Also in the MODELSEL screen.

**Comment by fraternl** [comment]
Yes, Taranis plus

**Comment by bsongis** [comment]
I reproduced only one time ... I changed trims and copied the current model to another slot. The radio hanged (not exactly the radio, but the menus task, the mixer is still alive, I am almost sure we are blocked inside an I2C wait).

**Comment by bsongis** [comment]
Now I reproduced 3 times ... I change trims and copy the current model to another slot in the same second or so. The radio hangs (not exactly the radio, but the menus task, the mixer is still alive).

The common function called in all our scenarios is eeCheck(true), pretty sure we are blocked somewhere inside, perhaps inside an I2C wait.

All ideas / other tests welcome, I am sure it's a firmware bug introduced in 2.1.0

**Comment by bsongis** [comment]
Now inside DELETE MODEL

**Comment by bsongis** [comment]
Still not good, I just had the problem on model delete

**Comment by bsongis** [comment]
In order to test my fix, I put an alert(...) just before I2C reinitialization to be aware I would be in an endless loop before (I mean without my fix). And yes I had the alert, the I2C was reinitialized and the write continued where it was. So I am confident, this issue is gone :)

**Comment by RC-SOAR** [comment]
That's great work - it sounded to me like it would be a tricky one to solve!

**Comment by projectkk2glider** [comment]
@bsongis Is this a I2C speed issue or is it caused by the fact that we now use hardware driver I2C?

**Comment by kilrah** [comment]
I would think it's related to the speed that might be borderline in some cases, that's the only thing I could think of that could cause problems...

**Comment by bsongis** [comment]
I am sure of nothing here, here are some possible causes:
- the EEPROM doesn't answer or SDA / SCL remain high => when I have more time, I need to check this with the logic analyser
- there is a timing problem, a task woken up during the I2C translation? It shouldn't be a problem, but ...
- perhaps there is an error detected by the I2C hardware layer, it is reported but we don't check it correctly 
- the STlib implementation of I2C is not perfect (many reports on the web for this)

**Comment by projectkk2glider** [comment]
I've created a new ticket to remind us of this remaining issue.
